runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java

```diff
@@ -0,0 +1,262 @@
+    List<MicrobatchSource<T, CheckpointMarkT>> result = new ArrayList<>();
+    List<? extends UnboundedSource<T, CheckpointMarkT>> splits =
+        source.generateInitialSplits(numInitialSplits, options);
+    int numSplits = splits.size();
+    long[] numRecords = splitNumRecords(maxNumRecords, numSplits);
+    for (int i = 0; i < numSplits; i++) {
+      // splits must be stable, and cannot change during consecutive executions
+      // for example: Kafka should not add partitions if more then one topic is read.
+      result.add(new MicrobatchSource<>(splits.get(i), maxReadTime, 1, numRecords[i], i));
+    }
+    return result;
+  }
+
+  @Override
+  public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+    return 0;
+  }
+
+  @Override
+  public boolean producesSortedKeys(PipelineOptions options) throws Exception {
+    return false;
+  }
+
+  @Override
+  public BoundedReader<T> createReader(PipelineOptions options) throws IOException {
+    return createReader(options, null);
+  }
+
+  public BoundedReader<T> createReader(PipelineOptions options, CheckpointMarkT checkpointMark)
+      throws IOException {
+    return new Reader(source.createReader(options, checkpointMark));
+  }
+
+  @Override
+  public void validate() {
+    source.validate();
+  }
+
+  @Override
+  public Coder<T> getDefaultOutputCoder() {
+    return source.getDefaultOutputCoder();
+  }
+
+  public Coder<CheckpointMarkT> getCheckpointMarkCoder() {
+    return source.getCheckpointMarkCoder();
+  }
+
+  public int getSplitId() {
+    return splitId;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof MicrobatchSource)) {
+      return false;
+    }
+    MicrobatchSource<?, ?> that = (MicrobatchSource<?, ?>) o;
+
+    return splitId == that.splitId;
+  }
+
+  @Override
+  public int hashCode() {
+    return splitId;
+  }
+
+  /**
+   * A {@link BoundedSource.BoundedReader}
+   * wrapping an {@link UnboundedSource.UnboundedReader}.
+   *
+   * <p>This Reader will read until it reached the bound of duration, or max records,
+   * whichever comes first.
+   */
+  public class Reader extends BoundedSource.BoundedReader<T> {
+    private long recordsRead = 0L;
+    private final Instant endTime;
+    private final FluentBackoff backoffFactory;
+    private final UnboundedSource.UnboundedReader<T> reader;
+
+    private Reader(UnboundedSource.UnboundedReader<T> reader) {
+      endTime = Instant.now().plus(maxReadTime);
+      this.reader = reader;
+      backoffFactory =
+          FluentBackoff.DEFAULT
+              .withInitialBackoff(Duration.millis(10))
+              .withMaxBackoff(maxReadTime.minus(1))
+              .withMaxCumulativeBackoff(maxReadTime.minus(1));
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      LOG.info("MicrobatchReader-{}: Starting a microbatch read from an unbounded source with a "
```

Comment:
Seems like `LOG.debug` might be more appropriate for this routine and frequent operation.